### PR TITLE
[nexus] migrate CoAP observe test to Nexus

### PR DESCRIPTION
**Pull request by jwhui**
This commit migrates the functionality covered by test_coap_observe.py to the Nexus test framework.

- Enabled OPENTHREAD_CONFIG_COAP_OBSERVE_API_ENABLE in Nexus config.
- Created test_coap_observe.cpp to test CoAP observations and notifications in a simulated network.
- Handled edge cases in the test to avoid segfaults during cancel response processing.
- Removed the old Python test test_coap_observe.py.

### CODE REVIEW

File: tests/nexus/test_coap_observe.cpp [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+    }
+
+    if (aResult != OT_ERROR_NONE)
+    {
+        return;
+    }
+
+    sNotificationReceived = true;
+
+    uint16_t length = otMessageGetLength(aMessage) - otMessageGetOffset(aMessage);
```
Comment (abtink) [inline tests/nexus/test_coap_observe.cpp]: This test is using `ot*` types and APIs.
We should avoid using these in nexus test implemenation and instead use the related C++ core types and APIs. Please update the test code to follow this. Thanks.

**Review by abtink — APPROVED** [review]
